## Using _openpyxl_ to Read from an Excel File

The _openpyxl_ package provides functionality for reading and writing Excel files in Python. There are other packages that provide similar functionality, but _openpyxl_ is conveniently included in the Anaconda Python distribution. First of all, let's import the _openpyxl_ package and use the shorter label _opx_ as an alias for the full package name. 

In [1]:
import openpyxl as opx

Next, let's open our Excel data file, _CRS_data.xlsx_, by using the _load_workbook()_ function. This function returns a workbook object which we'll assign to the variable name _wb_.

In [2]:
wb = opx.load_workbook('CRS_data.xlsx')

In [3]:
type(wb)

openpyxl.workbook.workbook.Workbook

In [4]:
#help(wb)

If you run the _help(wb)_ command, you will see there are a number of attributes and methods associated with a workbook object. We will only need to use a couple of them to read information from the workbook. The first attribute we'll use is _.sheetnames_ which stores a list of all the worksheet names in the workbook. There are six worksheets in the _CRS_data.xlsx_ workbook.

In [5]:
wb.sheetnames

['Forecasted Demand',
 'Annual Plant Capacity',
 'Inbound Freight Costs',
 'Handling Charges',
 'Outbound Freight Cost',
 'Transit Time']

We can use these sheet names like keys in a dictionary to select a specific worksheet object within the workbook. Let's use that syntax to select the 'Forecasted Demand' worksheet object in the _wb_ workbook and assign that worksheet object to the variable _ws_.

In [6]:
ws = wb['Forecasted Demand']

In [7]:
type(ws)

openpyxl.worksheet.worksheet.Worksheet

In [8]:
#help(ws)

Again, there are many attributes and methods associated with a worksheet object, but we'll only need to use the _.values_ attribute to retrieve the values stored in the cells of the worksheet object _ws_. The _.values_ attribute is actually a generator that returns a tuple for each row of values in the worksheet. We need to embed the _.values_ generator within another expression in order to retrieve those tuples. Let's create a list containing all rows of values in the spreadsheet and then display the first 10 tuples of rows.

In [9]:
lst = list(ws.values)
lst[0:10]

[('Customer Zone', 'Demand'),
 (10, 33081),
 (12, 18595),
 (14, 12046),
 (18, 65862),
 (19, 22170),
 (21, 139581),
 (22, 25917),
 (24, 17900),
 (27, 10691)]

Finally, we'll use a for loop to iterate over every row in the worksheet and use a tuple assignment statement to store each row of information in a dictionary using the customer zone numbers as the key values. We'll also want to delete the dictionary item containing the column headings from the dictionary.

In [10]:
ws = wb['Forecasted Demand']
demand = dict()
for row in ws.values:
    (zone, dmd) = row
    demand[zone] = dmd
del demand['Customer Zone']

We can now use the _demand_ dictionary to retrieve the forecasted demand value for any customer zone using the zone number as the dictionary key. For example, the forecasted demand for zone 19 is 22,170.

In [11]:
demand[19]

22170

You now have all the tools you need to retrieve the data from _CRS_data.xlsx_ to build a supply chain network optimization model for the Clifton Running Shoes case.